### ISSUE TITLE
[2.3.4] Assets controller throws NullPointerException on request for a directory

### ISSUE THREAD

**Issue by olafurpg** [playframework/playframework#3459]
Following a [discussion](https://groups.google.com/forum/#!topic/play-framework/aoFIsFjCnB0) from the mailing list.

1. Start a new play-scala app `activator new`, using 2.3.4
2. run the app on port 9000
3. make a GET request to http://localhost:9000/assets/images
4. app returns a 500 error with an uncaught NullPointerException on the server, full stacktrace: http://codepad.org/EUHRWLL2

I would have expected the asset controller to return a 400 Bad Request, 404 Not Found or something similar. I noticed this behaviour recently on a play app that we are using in production, we were attacked by some malicious user who performed almost 3k requests in a short period of time to various dubious URLs.

Example of this error found in the 'wild':

* https://launaskil.is/assets/images (our site)
* https://www.journiapp.com/assets/images (another site I found)

I am using play-scala, OSX 10.9, java version "1.7.0_67".

**Comment by EECOLOR** [comment]
Is there any progress on this issue?

**Comment by EECOLOR** [comment]
As noted by @jroper the actual problem is in the implementation of `JarURLInputStream` of `sun.net.www.protocol.jar.JarURLConnection`. It is unclear if this problem also occurs on OpenJDK.

How bad would it be to wrap the `close` call at [Assets.scala#L149](https://github.com/playframework/playframework/blob/2.3.x/framework/src/play/src/main/scala/play/api/controllers/Assets.scala#L149) in a try/catch for `NullPointerException`?

**Comment by jroper** [comment]
Wrapping the close would work - but is not ideal if the other suggestion I made is practical (I don't know if it is or not).

**Comment by eeggen-rhinofly** [comment]
Is it even an option to close the JarFile without closing the InputStream if the latter is not null? I guess not, because the close() method on JarURLInputStream *already* closes the JarFile after the input stream is closed. That's where the Exception is thrown.

**Comment by eeggen-rhinofly** [comment]
There seems to be no need for the `close` on the `InputStream`. The [commit that introduced this code](https://github.com/playframework/playframework/commit/3b78a4a184dd9904eaabd44426879e39da7b4aff) added the `try-finally` block. The original code:

``` scala
Option(resource.openConnection)		
  .map(_.asInstanceOf[JarURLConnection].getJarEntry.getTime)		
  .filterNot(_ == -1)
```
This implies that the code was not introduced as a bug fix.

On top of that, it does not make sense to `close` an input stream that was never opened. `getJarEntry.getTime` does not open the stream. The effect of the current implementation is that the stream is opened just to close it.

My suggestion would be to remove the `try-finally` block and the call to `close` the stream.

**Comment by jroper** [comment]
There is absolutely a resource leak in the original code there, if you open a connection to a resource, and that resource is a jar file, a file descriptor is created for accessing the jar.  If you don't close it, you leak a file descriptor.  It may eventually be cleaned up by a finalize method, but we should not rely on that.

**Comment by EECOLOR** [comment]
This is the code from `JarURLConnection.getInputStream`:

```
  149       public InputStream getInputStream() throws IOException {
  150           connect();
  151   
  152           InputStream result = null;
  153   
  154           if (entryName == null) {
  155               throw new IOException("no entry name specified");
  156           } else {
  157               if (jarEntry == null) {
  158                   throw new FileNotFoundException("JAR entry " + entryName +
  159                                                   " not found in " +
  160                                                   jarFile.getName());
  161               }
  162               result = new JarURLInputStream (jarFile.getInputStream(jarEntry));
  163           }
  164           return result;
  165       }
```

From a naive point of view that does not help in closing the connection to the jarfile. However, `JarURLInputStream` has an unexpected side effect in it's `close` method:

``` scala
  102       class JarURLInputStream extends java.io.FilterInputStream {
  103           JarURLInputStream (InputStream src) {
  104               super (src);
  105           }
  106           public void close () throws IOException {
  107               try {
  108                   super.close();
  109               } finally {
  110                   if (!getUseCaches()) {
  111                       jarFile.close();
  112                   }
  113               }
  114           }
  115       }
```

If it is not using caches, it closes the jar file. So if we want the exact same behaviour without creating an unneeded stream to an entry of the jar file we could use the following code:

``` scala
if (jarUrlConnection.getUseCaches) jarUrlConnection.getJarFile.close()
```

**Comment by jroper** [comment]
Yes, it's the `JarURLConnection` that holds the resource (file descriptor, socket etc) that needs to be closed - this is the case for all `URLConnection`'s.  The `InputStream` you get from it holds no resources, and it costs nothing to get it, but without casting the `URLConnection`, the only way to close a `URLConnection` is to get the input stream and close that, which is why it's done that way currently in the code.  We can (and should) implement the above code for `JarUrlConnection`, but we still need to do the `getInputStream.close` for all other types of url connections (the main other type here would be `FileUrlConnection`, which would be quite common for this code to encounter in dev mode) since we don't want to have to match on every possible type of `URLConnection` that could be returned and implement the method specific to that to close it.

I mentioned the need to check `getUseCases in my [original comment](https://github.com/playframework/playframework/issues/3459#issuecomment-57318827).

**Comment by sebdotv** [comment]
We're seeing those also on an internet facing server. Any chance for a 2.3.x fix/workaround?

**Comment by EECOLOR** [comment]
> Any chance for a 2.3.x fix/workaround?

**Comment by EECOLOR** [comment]
@jroper @pvlugter Any chance this will be backported to `2.3.x`?

**Comment by devsoftweb** [comment]
+1, would also like to see this on `2.3.x`

**Comment by kflorence** [comment]
+1 for backport to `2.3.x`

**Comment by alaz** [comment]
+1

**Comment by pkalmakis** [comment]
+1 for a backport. Is there a workaround we can use in the meantime? Can this exception be caught/suppressed?

**Comment by jroper** [comment]
This is fixed in Play 2.3.10.
